### DISCUSSION TITLE
Pulling non-Roman title from SRS or Inventory

### DISCUSSION THREAD

**Discussion by todolson** [folio-org/folio-analytics#494] [Discussions]
For our Pull List report, I need to pull the non-Roman title of a work, or null if there is nothing. In MARC terms, this is identifying the 880 (just ‡a & ‡b) that corresponds to the 245 of a record. Here's a record that illustrates the complexity (HRID 2491042 in my system):
```
001     2491042
100  1    ‡6 880-01  ‡a Chen, Guoxiong.  ‡0 http://id.loc.gov/authorities/names/n83017291
245  1 0  ‡6 880-02  ‡a Bie pa Zhongguo :  ‡b Dang qian Tai hai gong fang zhan liang an jun li zong bi jiao /  ‡c Chen Guoxiong zhu.
246  3 0  ‡6 880-03  ‡a Dang qian Tai hai gong fang zhan liang an jun li zong bi jiao
880  1    ‡6 100-01/$1  ‡a 陳國雄.  ‡0 http://id.loc.gov/authorities/names/n83017291
880  1 0  ‡6 245-02/$1  ‡a 别怕中國　:  ‡b 當前台海攻防戰兩岸軍力總比較 /  ‡c 陳國雄著.
880  3 0  ‡6 246-03/$1  ‡a 當前台海攻防戰兩岸軍力總比較
```
Stripping this down to relevant columns, this is stored in `srs_marctab` like so:
|instance_hrid|field|ord|sf|content|
|-------------|-----|---|--|-------|
|2491042|100|1|6|880-01|
|2491042|100|1|a|Chen, Guoxiong.|
|2491042|100|1|0|http://id.loc.gov/authorities/names/n83017291|
|2491042|245|1|6|880-02|
|2491042|245|1|a|Bie pa Zhongguo :|
|2491042|245|1|b|Dang qian Tai hai gong fang zhan liang an jun li zong bi jiao /|
|2491042|245|1|c|Chen Guoxiong zhu.|
|2491042|880|1|6|100-01/$1|
|2491042|880|1|a|陳國雄.|
|2491042|880|1|0|http://id.loc.gov/authorities/names/n83017291|
|2491042|880|2|6|245-02/$1|
|2491042|880|2|a|别怕中國　:|
|2491042|880|2|b|當前台海攻防戰兩岸軍力總比較 /|
|2491042|880|2|c|陳國雄著.|
|2491042|880|3|6|246-03/$1|
|2491042|880|3|a|當前台海攻防戰兩岸軍力總比較|

I _think_ what I want is (in mixed coding metaphor): 
for a given HRID, identify the 880 where `sf` = 6 and `content` matches `245%`, and return the 880 fields with the same `ord` where `sf` is in ("a", "b")

What I have in SQL is:
```sql
select array_to_string(array_agg("content"), ' ') 
from srs_marctab
where instance_hrid = '2491042'
and field = '880'
and sf in ('a', 'b')
and ord in (
  select ord 
  from srs_marctab
  where instance_hrid = '2491042'
  and field = '880' and sf = '6'
  and "content" like '245-%');
```
This seems like there must be a better solution, especially because of the need to reference the `instance_hrid` value in both the outer select and the sub-select. And this will need to become part of another query in an MS Access application.

Does anyone have any advice on how to improve this?

**Comment by amzoss** [comment]
I'm sure there are a million ways to do this, but I think this is at least logical:

```
with subset as (
     select
        srs_id,
        field,
        ord,
        sf,
        "content"
      from srs_marctab
      where field = '880' 
      and sf = '6'
      and "content" like '245-%'
)
select
    srs_marctab.srs_id,
    srs_marctab.instance_hrid,
    srs_marctab.field,
    srs_marctab.ord,
    srs_marctab.sf,
    srs_marctab."content"
from srs_marctab
right join subset
    on srs_marctab.srs_id = subset.srs_id
    and srs_marctab.field = subset.field
    and srs_marctab.ord = subset.ord
where srs_marctab.sf in ('a', 'b')
;
```

Basically, you can do the query for the `ord` without specifying an identifier, and then you can join those records back to the same table using the `identifier` and `field` and `ord` so you can grab the other `sf` values. And the first query really doesn't need all of those columns, but I found them helpful to double check.

Oh, and I have `srs_id` as the identifier here because snapshot doesn't have any data in instance_hrid, but you should be able to switch those out for each other.

**Reply by todolson** [reply]
Thank you, that works well.

First I tried the query as-is, limiting to a specific value for `srs_marctab.instance_h`, just to see how that worked. In that case, with the join on `srs_id` and the limit on `instance_hrid`, the query took a long time, I canceled after four minutes. So even though there's an index on `instance_hrid`, it didn't help when the join was on `srs_id`. 

Then I did as you suggested and changed the join to also use `instance_hrid`:
```sql
with subset as (
     select
        srs_id,
        instance_hrid,
        field,
        ord,
        sf,
        "content"
      from srs_marctab
      where field = '880' 
      and sf = '6'
      and "content" like '245-%'
)
select
    srs_marctab.srs_id,
    srs_marctab.instance_hrid,
    srs_marctab.field,
    srs_marctab.ord,
    srs_marctab.sf,
    srs_marctab."content"
from srs_marctab
right join subset
    on srs_marctab.instance_hrid = subset.instance_hrid
    and srs_marctab.field = subset.field
    and srs_marctab.ord = subset.ord
where srs_marctab.sf in ('a', 'b')
and srs_marctab.instance_hrid = '2491042';
```
And the response comes back in under a second! So there does seem to be something important about keeping the join and the limit in the same index, at least for this self-join.

**Reply by amzoss** [reply]
That's great to know, thanks!
